components/script/dom/windowproxy.rs

```diff
@@ -200,6 +204,134 @@ impl WindowProxy {
+                window_size: None,
+                layout_threads: PREFS.get("layout.threads").as_u64().expect("count") as usize,
+            };
+            let constellation_msg = ScriptMsg::ScriptNewAuxiliary(load_info, pipeline_sender);
+            window.send_to_constellation(constellation_msg);
+            ScriptThread::process_attach_layout(new_layout_info, document.origin().clone());
+            let msg = EmbedderMsg::BrowserCreated(new_top_level_browsing_context_id);
+            window.send_to_embedder(msg);
+            let auxiliary = ScriptThread::find_document(new_pipeline_id).and_then(|doc| doc.browsing_context());
+            if let Some(proxy) = auxiliary {
+                if name.to_lowercase() != "_blank" {
+                    proxy.set_name(name);
+                }
+                return Some(proxy)
+            }
+        }
+        None
+    }
+
+    // https://html.spec.whatwg.org/multipage/#window-open-steps
+    pub fn open(&self,
+                url: DOMString,
+                target: DOMString,
+                features: DOMString)
+                -> Option<DomRoot<WindowProxy>> {
+        // Step 3.
+        let non_empty_target = match target.as_ref() {
+            "" => DOMString::from("_blank"),
+            _ => target
+        };
+        // TODO Step 4, properly tokenize features.
+        let noopener = features.contains("noopener");
+        let (chosen, new)  = match self.choose_browsing_context(non_empty_target, noopener) {
+            (Some(window_proxy), new) => (window_proxy, new),
+            (None, _) => return None
+        };
+        if let Some(target_document) = chosen.document() {
+            let target_window = target_document.window();
+            // Step 9, and 10.2, will have happened elsewhere,
+            // since we've created a new browsing context and loaded it with about:blank.
+            if !url.is_empty() {
+                let existing_document = self.currently_active.get()
+                            .and_then(|id| ScriptThread::find_document(id)).unwrap();
+                // Step 10.1
+                let url = match existing_document.url().join(&url) {
+                    Ok(url) => url,
+                    Err(_) => return None, // TODO: throw a  "SyntaxError" DOMException.
+                };
+                // Step 10.3
+                target_window.load_url(url, new, false, target_document.get_referrer_policy());
+            }
+            if noopener {
+                // Step 11 (Dis-owning has been done in create_auxiliary_browsing_context).
+                return None
+            }
+            // Step 12.
+            return target_document.browsing_context()
+        };
+        None
+    }
+
+    // https://html.spec.whatwg.org/multipage/#the-rules-for-choosing-a-browsing-context-given-a-browsing-context-name
+    pub fn choose_browsing_context(&self, name: DOMString, noopener: bool) -> (Option<DomRoot<WindowProxy>>, bool) {
```

Comment:
Why the `Option<_>`?

Reply:
The only reason I can come up(I don't remember exactly why I did this), is that one arm of this one creates a new auxiliary, and per the spec if the user doesn't allow the pop up, we should 'do nothing'. So I guess the thing to do is return `Option` and let the calling code handle it? Should it be `Result` you think?

Reply:
`Option<_>` looks ok to me here.